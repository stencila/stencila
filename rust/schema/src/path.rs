use std::{
    collections::VecDeque,
    fmt::{self, Debug},
};

use derive_more::{Deref, DerefMut, IntoIterator};

use common::{
    eyre::{bail, OptionExt, Report, Result},
    serde::{Deserialize, Serialize},
    serde_json::{self},
};
use node_type::NodeProperty;

/// A slot in a node path: either a property identifier or the index of a vector.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(untagged, crate = "common::serde")]
pub enum NodeSlot {
    Property(NodeProperty),
    Index(usize),
}

impl From<NodeProperty> for NodeSlot {
    fn from(value: NodeProperty) -> Self {
        Self::Property(value)
    }
}

impl From<usize> for NodeSlot {
    fn from(value: usize) -> Self {
        Self::Index(value)
    }
}

impl TryFrom<serde_json::Value> for NodeSlot {
    type Error = Report;

    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
        use serde_json::Value::*;
        match value {
            String(name) => Ok(Self::from(NodeProperty::try_from(name.as_str())?)),
            Number(index) => index
                .as_u64()
                .ok_or_eyre("Expected u64")
                .map(|value| Self::from(value as usize)),
            _ => bail!("Unhandled JSON value for `PatchSlot`"),
        }
    }
}

/// A path to reach a node from the root: a vector of [`PatchSlot`]s
///
/// A [`VecDeque`], rather than a [`Vec`] so that when applying operations in
/// a call to `patch` the front of the path can be popped off.
#[derive(
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Deref,
    DerefMut,
    IntoIterator,
    Serialize,
    Deserialize,
)]
#[serde(crate = "common::serde")]
#[derive(Default)]
pub struct NodePath(VecDeque<NodeSlot>);

impl NodePath {
    pub fn new() -> Self {
        Self::default()
    }
}

impl From<NodeProperty> for NodePath {
    fn from(value: NodeProperty) -> Self {
        Self::from(NodeSlot::from(value))
    }
}

impl From<usize> for NodePath {
    fn from(value: usize) -> Self {
        Self::from(NodeSlot::from(value))
    }
}

impl From<NodeSlot> for NodePath {
    fn from(value: NodeSlot) -> Self {
        Self(VecDeque::from([value]))
    }
}

impl<const N: usize> From<[NodeSlot; N]> for NodePath {
    fn from(value: [NodeSlot; N]) -> Self {
        Self(value.into())
    }
}

impl<const N: usize> From<[NodeProperty; N]> for NodePath {
    fn from(value: [NodeProperty; N]) -> Self {
        Self(value.into_iter().map(NodeSlot::from).collect())
    }
}

impl TryFrom<serde_json::Value> for NodePath {
    type Error = Report;

    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
        use serde_json::Value::*;
        match value {
            Number(..) | String(..) => Ok(Self::from(NodeSlot::try_from(value)?)),
            Array(array) => Ok(Self(VecDeque::from(
                array
                    .into_iter()
                    .map(NodeSlot::try_from)
                    .collect::<Result<Vec<_>>>()?,
            ))),
            _ => bail!("Unhandled JSON value for `PatchPath`"),
        }
    }
}

impl Debug for NodePath {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (index, slot) in self.iter().enumerate() {
            if index != 0 {
                f.write_str(".")?;
            }
            slot.fmt(f)?;
        }

        Ok(())
    }
}
