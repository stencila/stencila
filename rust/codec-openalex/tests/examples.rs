use std::path::PathBuf;

use glob::glob;
use insta::assert_json_snapshot;

use stencila_codec::{
    Codec,
    eyre::{Context, OptionExt, Result, eyre},
};
use stencila_codec_openalex::OpenAlexCodec;

/// Decode CSL-JSON into articles
#[tokio::test]
async fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*.json";

    for path in glob(&pattern)?.flatten() {
        let id = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(".json").map(String::from))
            .ok_or_eyre("should have .json suffix")?;

        let (node, ..) = OpenAlexCodec
            .from_path(&path, None)
            .await
            .wrap_err_with(|| eyre!("Unable to deserialize {id}.json"))?;

        assert_json_snapshot!(id, node, {
            ".commit" => "redacted"
        });
    }

    Ok(())
}
