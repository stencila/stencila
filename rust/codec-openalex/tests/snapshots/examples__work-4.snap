---
source: rust/codec-openalex/tests/examples.rs
expression: node
---
{
  "type": "Article",
  "doi": "10.1016/s0021-9258(19)52451-6",
  "abstract": [
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Since 1922 when Wu proposed the use of the Folin phenol reagent for the measurement of proteins (l), a number of modified analytical procedures ut.ilizing this reagent have been reported for the determination of proteins in serum (2-G), in antigen-antibody precipitates (7-9), and in insulin (10).Although the reagent would seem to be recommended by its great sensitivity and the simplicity of procedure possible with its use, it has not found great favor for general biochemical purposes.In the belief that this reagent, nevertheless, has considerable merit for certain application, but that its peculiarities and limitations need to be understood for its fullest exploitation, it has been studied with regard t.o effects of variations in pH, time of reaction, and concentration of reactants, permissible levels of reagents commonly used in handling proteins, and interfering subst.ances.Procedures are described for measuring protein in solution or after precipitation wit,h acids or other agents, and for the determination of as little as 0.2 y of protein. MethodReagents-Reagent A, 2 per cent N&OX in 0.10 N NaOH.Reagent B, 0.5 per cent CuS04.5Hz0 in 1 per cent sodium or potassium tartrabe.Reagent C, alkaline copper solution.Mix 50 ml. of Reagent A with 1 ml. of Reagent B. Discard after 1 day.Reagent D, carbonate-copper solution, is the same as Reagent C except for omission of NaOH.Reagent E, diluted Folin reagent.Titrate Folin-Ciocalteu phenol reagent ((II), Eimer and Amend, Fisher Scientific Company, New York) with NaOH t.o a phenolphthalein end-point.On the basis of this titration dilute the Folin reagent (about 2-fold) to make it 1 N in acid.Working standards may be prepared from human serum diluted IOO-to lOOO-fold (approximately 700 to 70 y per ml.).These in turn may be checked against a standard solution of crystalline bovine albumin (Armour and"
          }
        }
      ]
    }
  ],
  "authors": [
    {
      "type": "Person",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Washington University in St. Louis",
          "ror": "01yc7t268",
          "identifiers": [
            "https://openalex.org/I204465549"
          ]
        }
      ],
      "identifiers": [
        "https://openalex.org/A5110127684"
      ],
      "name": "OliverH. Lowry"
    },
    {
      "type": "Person",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Washington University in St. Louis",
          "ror": "01yc7t268",
          "identifiers": [
            "https://openalex.org/I204465549"
          ]
        }
      ],
      "identifiers": [
        "https://openalex.org/A5032482932"
      ],
      "name": "NiraJ. Rosebrough"
    },
    {
      "type": "Person",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Washington University in St. Louis",
          "ror": "01yc7t268",
          "identifiers": [
            "https://openalex.org/I204465549"
          ]
        }
      ],
      "familyNames": [
        "Farr"
      ],
      "identifiers": [
        "https://openalex.org/A5004071084"
      ],
      "name": "A. Farr"
    },
    {
      "type": "Person",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Washington University in St. Louis",
          "ror": "01yc7t268",
          "identifiers": [
            "https://openalex.org/I204465549"
          ]
        }
      ],
      "identifiers": [
        "https://openalex.org/A5109894513"
      ],
      "name": "RoseJ. Randall"
    }
  ],
  "datePublished": {
    "type": "Date",
    "value": "1951-11-01"
  },
  "title": [
    {
      "type": "Text",
      "value": {
        "string": "PROTEIN MEASUREMENT WITH THE FOLIN PHENOL REAGENT"
      }
    }
  ],
  "content": [],
  "identifiers": [
    "https://openalex.org/W1775749144",
    "https://doi.org/10.1016/s0021-9258(19)52451-6",
    {
      "type": "PropertyValue",
      "propertyId": "mag",
      "value": "1775749144"
    },
    "https://pubmed.ncbi.nlm.nih.gov/14907713"
  ],
  "isPartOf": {
    "type": "PublicationIssue",
    "isPartOf": {
      "type": "PublicationVolume",
      "isPartOf": {
        "type": "Periodical",
        "name": "Journal of Biological Chemistry"
      },
      "volumeNumber": 193
    },
    "issueNumber": 1
  },
  "licenses": [
    "https://creativecommons.org/licenses/by/4.0/"
  ],
  "repository": "https://github.com/stencila/stencila",
  "path": "rust/codec-openalex/tests/examples/work-4.json",
  "commit": "redacted",
  "pageStart": 265,
  "pageEnd": 275
}
